diff -urN pxview-0.2.3.org/src/main.c pxview-0.2.3/src/main.c
--- pxview-0.2.3.org/src/main.c	2004-12-23 10:20:02.000000000 +0100
+++ pxview-0.2.3/src/main.c	2005-04-01 11:30:07.857552184 +0200
@@ -2164,11 +2164,26 @@
 							first = 1;
 							break;
 					}
-					if(i < pxh->px_primarykeyfields)
-						fprintf(outfp, " unique");
 				}
 				pxf++;
 			}
+			if(pxh->px_primarykeyfields) {
+				first = 0;
+				pxf = pxh->px_fields;
+				fprintf(outfp, ",\n  unique(");
+				for(i=0; i<pxh->px_primarykeyfields; i++) {
+					if(fieldregex == NULL ||  selectedfields[i]) {
+						strrep(pxf->px_fname, ' ', '_');
+						if(first == 1)
+							fprintf(outfp, ",");
+						if(i < pxh->px_primarykeyfields)
+							fprintf(outfp, "%s", pxf->px_fname);
+						first = 1;
+					}
+					pxf++;
+				}
+				fprintf(outfp, ")");
+			}
 			fprintf(outfp, "\n);\n");
 
 			/* Create the indexes */
